cmd/zed/agents/zfs_mod.c

```diff
@@ -0,0 +1,911 @@
+		return;
+
+	(void) nvlist_lookup_string(vdev, ZPOOL_CONFIG_PHYS_PATH, &physpath);
+	(void) nvlist_lookup_uint64(vdev, ZPOOL_CONFIG_WHOLE_DISK, &wholedisk);
+	(void) nvlist_lookup_uint64(vdev, ZPOOL_CONFIG_OFFLINE, &offline);
+	(void) nvlist_lookup_uint64(vdev, ZPOOL_CONFIG_GUID, &guid);
+
+	if (offline)
+		return;  /* don't intervene if it was taken offline */
+
+	zed_log_msg(LOG_INFO, "zfs_process_add: pool '%s' vdev '%s' (%llu)",
+	    zpool_get_name(zhp), path, (long long unsigned int)guid);
+
+	/*
+	 * The VDEV guid is preferred for identification (gets passed in path)
+	 */
+	if (guid != 0) {
+		(void) snprintf(fullpath, sizeof (fullpath), "%llu",
+		    (long long unsigned int)guid);
+	} else {
+		/*
+		 * otherwise use path sans partition suffix for whole disks
+		 */
+		(void) strlcpy(fullpath, path, sizeof (fullpath));
+		if (wholedisk)
+			zfs_strip_partition(fullpath);
+	}
+
+	/*
+	 * Attempt to online the device.
+	 */
+	if (zpool_vdev_online(zhp, fullpath,
+	    ZFS_ONLINE_CHECKREMOVE | ZFS_ONLINE_UNSPARE, &newstate) == 0 &&
+	    (newstate == VDEV_STATE_HEALTHY ||
+	    newstate == VDEV_STATE_DEGRADED)) {
+		zed_log_msg(LOG_INFO, "  zpool_vdev_online: vdev %s is %s",
+		    fullpath, (newstate == VDEV_STATE_HEALTHY) ?
+		    "HEALTHY" : "DEGRADED");
+		return;
+	}
+
+	/*
+	 * If the pool doesn't have the autoreplace property set, then attempt
+	 * a true online (without the unspare flag), which will trigger a FMA
+	 * fault.
+	 */
+	if (!zpool_get_prop_int(zhp, ZPOOL_PROP_AUTOREPLACE, NULL) ||
+	    !wholedisk || physpath == NULL) {
+		(void) zpool_vdev_online(zhp, fullpath, ZFS_ONLINE_FORCEFAULT,
+		    &newstate);
+		zed_log_msg(LOG_INFO, "  zpool_vdev_online: %s FORCEFAULT (%s)",
+		    fullpath, libzfs_error_description(g_zfshdl));
+		return;
+	}
+
+	/*
+	 * convert physical path into its current device node
+	 */
+	(void) snprintf(rawpath, sizeof (rawpath), "%s%s", DEV_BYPATH_PATH,
+	    physpath);
+	if (realpath(rawpath, devpath) == NULL) {
+		zed_log_msg(LOG_INFO, "  realpath: %s failed (%s)",
+		    rawpath, strerror(errno));
+
+		(void) zpool_vdev_online(zhp, fullpath, ZFS_ONLINE_FORCEFAULT,
+		    &newstate);
+
+		zed_log_msg(LOG_INFO, "  zpool_vdev_online: %s FORCEFAULT (%s)",
+		    fullpath, libzfs_error_description(g_zfshdl));
+		return;
+	}
+
+	/*
+	 * we're auto-replacing a raw disk, so label it first
+	 */
+	if (!labeled) {
+		char *leafname;
+
+		/*
+		 * If this is a request to label a whole disk, then attempt to
+		 * write out the label.  Before we can label the disk, we need
+		 * to map the physical string that was matched on to the under
+		 * lying device node.
+		 *
+		 * If any part of this process fails, then do a force online
+		 * to trigger a ZFS fault for the device (and any hot spare
+		 * replacement).
+		 */
+#if 0
+		fd = open(devpath, O_RDWR|O_EXCL);
+		if (fd == -1) {
+			if (errno == EBUSY)
+				return;
+		} else {
+			(void) close(fd);
+		}
+#endif
```

Comment:
Can this hunk simply be removed?  We could probably safely add the O_EXCL flag to `zpool_label_disk`.